theano/gof/wrapper.py

```diff
@@ -0,0 +1,446 @@
+        void extract(PyObject* object, int field_pos) {
+            switch(field_pos) {
+                // Extraction cases.
+                %s
+                // Default case.
+                default:
+                    PyErr_Format(PyExc_TypeError, "Wrapper: no extraction defined for a field %%d.", field_pos);
+                    this->setErrorOccurred();
+                    this->cleanup();
+                    break;
+            }
+        }
+        """ % ('\n'.join(
+            [('case %d: extract_%s(object); break;' % (i, self.fields[i])) for i in range(self.length)])
+        )
+        return """
+        #ifndef %(struct_name_defined)s
+        #define %(struct_name_defined)s
+        struct %(struct_name)s {
+            /* Attributes, */
+            int %(struct_name)s_error;
+            %(struct_fields)s
+
+            /* Constructor. */
+            %(struct_name)s() {
+                %(struct_name)s_error = 0;
+                %(struct_init)s
+            }
+
+            /* Destructor. */
+            ~%(struct_name)s() {
+                // cleanup() is defined below.
+                cleanup();
+            }
+
+            /* Cleanup method. */
+            void cleanup() {
+                %(struct_cleanup)s
+            }
+
+            /* Extraction methods. */
+            %(struct_extraction_methods)s
+
+            /* Extract method. */
+            %(struct_extract_method)s
+
+            /* Other methods. */
+            void setErrorOccurred() {
+                ++%(struct_name)s_error;
+            }
+            int errorOccurred() {
+                return %(struct_name)s_error;
+            }
+        };
+        #endif
+        """ % locals()
+
+    def c_code_cache_version(self):
+        return (1, 1)
+
+    def c_declare(self, name, sub, check_input=True):
+        struct_name = self.name
+        return """
+        %(struct_name)s %(name)s;
+        """ % locals()
+
+    # c_init() and c_cleanup() are useless if we create the struct on stack
+    # because the struct has constructor and destructor.
+
+    def c_init(self, name, sub):
+        return ""
+
+    def c_cleanup(self, name, sub):
+        return ""
+
+    def c_extract(self, name, sub, check_input=True):
+        fail = sub['fail']
+        length = self.length
+        fields_list = '"%s"' % '", "'.join(self.fields)
+        check = 1 if check_input else 0
+        return """
+        const char* fields[] = {%(fields_list)s};
+        if (%(check)s) {
+            if (py_%(name)s == Py_None) {
+                PyErr_SetString(PyExc_ValueError, "Wrapper: expected an object, not None.");
+                %(fail)s
+            }
+            for (int i = 0; i < %(length)s; ++i) {
+                if (!PyObject_HasAttrString(py_%(name)s, fields[i])) {
+                    PyErr_Format(PyExc_TypeError, "Wrapper: missing expected attribute %%s in object.", fields[i]);
+                    %(fail)s
+                }
+            }
```

Comment:
No need to do that check since GetAttrString will fail below when the attribute is not present anyway.